docs/developer-docs/integrations/sns/get-sns/testflight.md

```diff
@@ -0,0 +1,258 @@
+export PEM_FILE="/home/martin/.config/dfx/identity/$(dfx identity whoami)/identity.pem"
+export CID="$(dfx canister id test)"
+sns-quill --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal --proposal "(record { title=\"Register dapp's canisters with SNS.\"; url=\"https://example.com/\"; summary=\"This proposal registers dapp's canisters with SNS.\"; action=opt variant {RegisterDappCanisters = record {canister_ids=vec {principal\"$CID\"}}}})" $DEVELOPER_NEURON_ID > register.json
+sns-quill send register.json
+```
+When running the testflight on mainnet, pass `--network ic` as an additional argument to `dfx canister`
+when obtaining the dapp's canister IDs.
+
+You can also register multiple canisters via a single SNS proposals by adjusting `RegisterDappCanisters`
+in the `--proposal` argument above to, e.g.,
+```
+RegisterDappCanisters = record {canister_ids=vec {principal\"$CID1\"; principal\"$CID2\";}}
+```
+
+Once the SNS proposal is executed, you should see all the registered dapp's canisters listed as dapps in
+```
+dfx canister call sns_root list_sns_canisters '(record {})'
+```
+When running the testflight on mainnet, pass `--network ic` as an additional argument to `dfx canister` above.
+
+#### 6. Test upgrading canisters via SNS proposals.
+
+Determine the path to the new WASM binary that you want to upgrade the canister to.
+For projects build with `dfx`, this binary is typically located at
+`.dfx/<network>/canisters/<canister-name>/<canister-name>.wasm`
+under the root directory of your project,
+where `<network>` is the network (e.g., `local` or `ic`) and `<canister-name>` is the name
+of the canister according to `dfx.json`.
+
+Now you can prepare and send the SNS proposal via `sns-quill`:
+```
+sns-quill --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-upgrade-canister-proposal --summary "This proposal upgrades test canister." --title "Upgrade test canister." --url "https://example.com/" --target-canister-id $CID --wasm-path "./.dfx/local/canisters/test/test.wasm" $DEVELOPER_NEURON_ID > upgrade.json
+sns-quill send upgrade.json | grep -v "^ *new_canister_wasm"
+```
+
+You can omit `grep -v "^ *new_canister_wasm"` above to see the new WASM binary in the output. Note that the output then contains the entire WASM binary and can be huge!
+
+#### 7. Test executing code on SNS managed canisters via SNS proposals.
+
+To execute code on SNS managed canisters via SNS proposals,
+the canisters must expose a pair of public
+functions (refered to as *generic* functions in the following):
+- a validation function to validate and render the proposal payload;
+- an execution function to perform an action given the proposal payload.
+
+The validation function must return `Result<String, String>`.
+If the validation function returns `Ok(rendering)`, then
+the proposal is submitted and the `rendering` string is included
+into the proposal.
+Otherwise, the proposal submission fails.
```

Comment:
Those are Rust types, which is fine, but I'd like to also see Candid types as that is the IDL we use in general.

Reply:
Not quite sure what the return type must look like as Candid. The code decodes the payload as follows:
```
let result = Decode!(&reply, Result<String,String>);
```

Reply:
I think it would be `variant { Ok: text; Err: text; }` if we take the [ledger's .did](https://github.com/dfinity/ic/blob/master/rs/rosetta-api/icp_ledger/ledger.did#L76-L79) as an example

Reply:
Thank you! Done